File: src/Butil/Bit.Butil/Publics/ViewTransition.cs
using System;
using System.Collections.Concurrent;
using System.Diagnostics.CodeAnalysis;
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace Bit.Butil;

/// <summary>
/// Wraps the <see href="https://developer.mozilla.org/en-US/docs/Web/API/View_Transition_API">View Transition API</see>:
/// the browser snapshots the page, lets you change it, and animates between the two states - so a
/// list reordering or a detail panel opening cross-fades instead of snapping.
/// </summary>
/// <remarks>
/// The animation is CSS. Butil starts and controls the transition; which elements morph into which
/// is decided by giving them a matching <c>view-transition-name</c>, and the default cross-fade
/// needs no CSS at all.
/// <br/>
/// <b>The timing that matters:</b> the browser takes the "before" snapshot when the transition
/// starts, calls your update callback, and takes the "after" snapshot when that callback's task
/// completes. In Blazor the DOM changes on a render, not on an assignment - so a callback that
/// merely sets a field and returns has not changed anything yet, and both snapshots come out the
/// same. Await the render before returning; the demo page shows the
/// <see cref="TaskCompletionSource"/>-in-<c>OnAfterRenderAsync</c> pattern that does this reliably.
/// </remarks>
[ButilService(typeof(ViewTransition))]
public class ViewTransition(IJSRuntime js) : IAsyncDisposable
{
    internal const string UpdateMethodName = nameof(InvokeViewTransitionUpdate);
    internal const string PhaseMethodName = nameof(InvokeViewTransitionPhase);
    internal const string PageMethodName = nameof(InvokeCrossDocumentEvent);

    private readonly ConcurrentDictionary<Guid, Func<Task>> _updates = new();
    private readonly ConcurrentDictionary<Guid, ViewTransitionHandle> _handles = new();
    private readonly ConcurrentDictionary<Guid, Action<CrossDocumentTransitionEvent>> _pageHandlers = new();

    // Per-instance callback reference (see Keyboard): transitions are isolated per circuit / WASM
    // app and released on disposal - no static state, no cross-circuit leak.
    private DotNetObjectReference<ViewTransition>? _dotNetRef;
    private DotNetObjectReference<ViewTransition> DotNetRef => DotNetObjectReferenceHelper.GetOrCreate(ref _dotNetRef, this);

    /// <summary>True when the runtime exposes <c>document.startViewTransition</c>.</summary>
    /// <remarks>
    /// Treat view transitions as a progressive enhancement: when this is false, do the same DOM
    /// change without one and the page still works, just without the animation.
    /// <br/>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> IsSupported() => js.Invoke<bool>("BitButil.viewTransition.isSupported");

    /// <summary>
    /// Invoked from JS to run the caller's DOM update between the two snapshots. Public +
    /// <see cref="JSInvokableAttribute"/> so it can be dispatched through the per-instance
    /// <see cref="DotNetObjectReference{T}"/>.
    /// </summary>
    /// <remarks>
    /// The returned task is what the browser awaits before snapshotting the new state, so this
    /// must not complete before the DOM has actually changed.
    /// </remarks>
    [JSInvokable(UpdateMethodName)]
    public async Task InvokeViewTransitionUpdate(Guid id)
    {
        if (_updates.TryGetValue(id, out var update)) await update();
    }

    /// <summary>
    /// Invoked from JS as the transition moves through its phases. Public +
    /// <see cref="JSInvokableAttribute"/> so it can be dispatched through the per-instance
    /// <see cref="DotNetObjectReference{T}"/>.
    /// </summary>
    [JSInvokable(PhaseMethodName)]
    public void InvokeViewTransitionPhase(Guid id, string phase, string message)
    {
        if (_handles.TryGetValue(id, out var handle) is false) return;

        handle.Advance(phase, message);

        // 'ready' and 'skipped' are mid-flight; only the terminal phases retire the handle.
        if (phase is "finished" or "failed")
        {
            _handles.TryRemove(id, out _);
            _updates.TryRemove(id, out _);
        }
    }

    /// <summary>
    /// Runs <paramref name="updateDom"/> inside a view transition, animating from the page as it
    /// looks now to the page as it looks afterwards.
    /// </summary>
    /// <param name="updateDom">
    /// Makes the change. It must not return until the DOM has actually updated - in Blazor that
    /// means awaiting a render, not just assigning a field.
    /// </param>
    /// <param name="types">
    /// Optional <see href="https://developer.mozilla.org/en-US/docs/Web/API/ViewTransition/types">transition types</see>,
    /// which CSS can select on to animate a "forward" navigation differently from a "back" one.
    /// Ignored by engines that only implement level 1.
    /// </param>
    /// <returns>
    /// A handle for awaiting the animation and skipping it, or null when the engine has no view
    /// transitions. On null, <paramref name="updateDom"/> is <b>not</b> run - the caller does the
    /// same change unanimated, which keeps the "did the update happen" question in one place.
    /// </returns>
    public async ValueTask<ViewTransitionHandle?> Start(Func<Task> updateDom, string[]? types = null)
    {
        ArgumentNullException.ThrowIfNull(updateDom);

        var id = Guid.NewGuid();
        var handle = new ViewTransitionHandle(js, id);

        _updates[id] = updateDom;
        _handles[id] = handle;

        var started = await js.Invoke<bool>("BitButil.viewTransition.start", DotNetRef, id, types);
        if (started) return handle;

        _updates.TryRemove(id, out _);
        _handles.TryRemove(id, out _);
        return null;
    }

    /// <summary>
    /// The common case: change some state and re-render inside a transition.
    /// </summary>
    /// <param name="updateState">Applies the state change. Runs before the re-render is requested.</param>
    /// <param name="render">
    /// Renders and completes once the DOM has caught up - pass a delegate that awaits your
    /// component's next render pass.
    /// </param>
    /// <param name="types">Optional transition types - see <see cref="Start(Func{Task}, string[])"/>.</param>
    /// <returns>The handle, or null when view transitions are unavailable. On null, the caller
    /// should apply the change itself without one.</returns>
    /// <remarks>
    /// This is <see cref="Start(Func{Task}, string[])"/> with the two halves named, so it is
    /// obvious that the render is the part that has to be awaited.
    /// </remarks>
    public ValueTask<ViewTransitionHandle?> Start(Action updateState, Func<Task> render, string[]? types = null)
    {
        ArgumentNullException.ThrowIfNull(updateState);
        ArgumentNullException.ThrowIfNull(render);

        return Start(async () =>
        {
            updateState();
            await render();
        }, types);
    }

    /// <summary>
    /// True when the runtime implements <b>cross-document</b> view transitions (View Transitions
    /// level 2) - the <c>@view-transition</c> opt-in plus the <c>pageswap</c>/<c>pagereveal</c> events.
    /// </summary>
    /// <remarks>
    /// Independent of <see cref="IsSupported"/>: an engine can animate within a document without
    /// being able to animate across a navigation.
    /// <br/>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> IsCrossDocumentSupported() => js.Invoke<bool>("BitButil.viewTransition.isCrossDocumentSupported");

    /// <summary>True when this document has been opted in through <see cref="EnableCrossDocument"/>.</summary>
    public ValueTask<bool> IsCrossDocumentEnabled() => js.Invoke<bool>("BitButil.viewTransition.isCrossDocumentEnabled");

    /// <summary>
    /// Opts this document into animating across same-origin navigations, by installing the
    /// <c>@view-transition { navigation: auto; }</c> rule.
    /// </summary>
    /// <param name="types">
    /// Optional transition types the incoming document's CSS can select on - the level-2 way to
    /// animate a "forward" navigation differently from a "back" one.
    /// </param>
    /// <remarks>
    /// <b>Both documents have to opt in</b> - the one being left and the one being entered - and the
    /// navigation has to be same-origin. There is no scripted switch for this in the spec: the
    /// at-rule is the whole opt-in, so this installs it as a stylesheet rather than calling anything.
    /// <br/>
    /// In a Blazor app this only applies to navigations that actually leave the document. Blazor's
    /// own router navigates without one, so an in-app route change is a job for
    /// <see cref="Start(Func{Task}, string[])"/> instead.
    /// </remarks>
    public ValueTask<bool> EnableCrossDocument(string[]? types = null)
        => js.Invoke<bool>("BitButil.viewTransition.enableCrossDocument", types);

    /// <summary>Removes the opt-in, so navigations stop animating.</summary>
    public ValueTask DisableCrossDocument() => js.InvokeVoid("BitButil.viewTransition.disableCrossDocument");

    /// <summary>
    /// Invoked from JS for a page swap or reveal. Public + <see cref="JSInvokableAttribute"/> so it
    /// can be dispatched through the per-instance <see cref="DotNetObjectReference{T}"/>.
    /// </summary>
    [JSInvokable(PageMethodName)]
    public void InvokeCrossDocumentEvent(Guid id, CrossDocumentTransitionEvent transitionEvent)
    {
        if (_pageHandlers.TryGetValue(id, out var handler)) handler.Invoke(transitionEvent ?? new CrossDocumentTransitionEvent());
    }

    /// <summary>
    /// Fires on the <b>outgoing</b> document just before the browser snapshots it - the last chance
    /// to set up what the transition animates away from.
    /// </summary>
    /// <returns>A subscription that detaches the listener on dispose.</returns>
    /// <remarks>
    /// The document is on its way out, so anything asynchronous started here is racing its own
    /// teardown. Keep the handler synchronous.
    /// </remarks>
    [DynamicDependency(nameof(InvokeCrossDocumentEvent), typeof(ViewTransition))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(CrossDocumentTransitionEvent))]
    public Task<ButilSubscription> OnPageSwap(Action<CrossDocumentTransitionEvent> handler)
        => SubscribePageEvent(handler, "pageswap");

    /// <summary>
    /// Fires on the <b>incoming</b> document just before its first paint - where the arriving page
    /// decides how the transition should look, or skips it.
    /// </summary>
    /// <returns>A subscription that detaches the listener on dispose.</returns>
    [DynamicDependency(nameof(InvokeCrossDocumentEvent), typeof(ViewTransition))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(CrossDocumentTransitionEvent))]
    public Task<ButilSubscription> OnPageReveal(Action<CrossDocumentTransitionEvent> handler)
        => SubscribePageEvent(handler, "pagereveal");

    private async Task<ButilSubscription> SubscribePageEvent(Action<CrossDocumentTransitionEvent> handler, string name)
    {
        ArgumentNullException.ThrowIfNull(handler);

        var id = Guid.NewGuid();
        _pageHandlers[id] = handler;

        await js.Invoke<bool>("BitButil.viewTransition.onPageEvent", DotNetRef, id, PageMethodName, name);

        return new ButilSubscription(id, async () =>
        {
            _pageHandlers.TryRemove(id, out _);
            await js.InvokeVoid("BitButil.viewTransition.offPageEvent", id);
        });
    }

    /// <summary>
    /// On scope/circuit teardown, skips any transition still running so a half-finished animation
    /// can't outlive the page that started it.
    /// </summary>
    public async ValueTask DisposeAsync()
    {
        try
        {
            _updates.Clear();
            _handles.Clear();
            _pageHandlers.Clear();
            await js.InvokeVoid("BitButil.viewTransition.disposeAll");
        }
        catch (Exception ex) when (ex.IsIgnorableDisposalException()) { } // teardown: circuit gone, cancelled, or already disposed
        finally
        {
            _dotNetRef?.Dispose();
            _dotNetRef = null;
        }
        GC.SuppressFinalize(this);
    }
}

File: src/Butil/Bit.Butil/Scripts/viewTransition.ts
var BitButil = (window as any).BitButil = (window as any).BitButil || {};

(function (butil: any) {
    const _transitions: { [id: string]: any } = {};
    const _pageListeners: { [id: string]: { name: string, handler: any } } = {};

    // The stylesheet that opts this document into cross-document transitions. Level 2 has no
    // scripted switch - `@view-transition { navigation: auto }` is the whole opt-in - so enabling it
    // from C# means injecting the rule and removing it again.
    const OPT_IN_ID = 'bit-butil-view-transition-opt-in';

    function describePageEvent(e: any, name: string) {
        const activation = e?.activation;
        return {
            name,
            hasTransition: !!e?.viewTransition,
            fromUrl: activation?.from?.url ?? '',
            toUrl: activation?.entry?.url ?? '',
            navigationType: activation?.navigationType ?? ''
        };
    }

    // 'pagereveal' fires on the incoming document before it is first painted - before .NET has
    // started, let alone subscribed. So the listener goes on while this module is evaluated (which,
    // for a bundle on the page, is early enough) and the event is parked until a subscriber asks for
    // it, the way launchQueue parks a launch. Under lazy scripts the module is imported on the first
    // call and there is nothing to park - the event is simply gone by then.
    let _revealed: any = null;
    if ('onpagereveal' in window) {
        window.addEventListener('pagereveal', (e: any) => { _revealed = describePageEvent(e, 'pagereveal'); }, { once: true });
    }

    butil.viewTransition = {
        isSupported() { return typeof (document as any).startViewTransition === 'function'; },
        isCrossDocumentSupported() {
            // Both halves have to be there: the opt-in rule the browser understands, and the events
            // that let a page take part in the transition it triggers.
            return 'CSSViewTransitionRule' in window && 'onpageswap' in window;
        },
        isCrossDocumentEnabled() { return !!document.getElementById(OPT_IN_ID); },
        enableCrossDocument(types: string[] | null) {
            let style = document.getElementById(OPT_IN_ID) as HTMLStyleElement | null;
            if (!style) {
                style = document.createElement('style');
                style.id = OPT_IN_ID;
                document.head.appendChild(style);
            }
            const typeRule = types?.length ? `\n  types: ${types.join(' ')};` : '';
            style.textContent = `@view-transition {\n  navigation: auto;${typeRule}\n}`;
            return true;
        },
        disableCrossDocument() {
            document.getElementById(OPT_IN_ID)?.remove();
        },
        onPageEvent(dotNetRef: any, listenerId: string, method: string, name: string) {
            // 'pageswap' fires on the old document just before it is snapshotted; 'pagereveal' on
            // the new one just before it is painted. They are the two places a cross-document
            // transition can be customized - or opted out of, by skipping it.
            if (!(`on${name}` in window)) return false;

            // Re-registering the same listener id would otherwise orphan the previous handler.
            butil.viewTransition.offPageEvent(listenerId);

            const handler = (e: any) => butil.utils.dispatch(dotNetRef, method, listenerId, describePageEvent(e, name));

            window.addEventListener(name, handler);
            _pageListeners[listenerId] = { name, handler };

            // This document's own 'pagereveal' has already happened; replaying it is the only way a
            // subscriber can see how this document was navigated into.
            if (name === 'pagereveal' && _revealed) {
                const revealed = _revealed;
                _revealed = null;
                butil.utils.dispatch(dotNetRef, method, listenerId, revealed);
            }

            return true;
        },
        offPageEvent(listenerId: string) {
            const entry = _pageListeners[listenerId];
            if (!entry) return;
            delete _pageListeners[listenerId];
            try { window.removeEventListener(entry.name, entry.handler); } catch { /* window gone */ }
        },
        start(dotNetRef: any, id: string, types: string[] | null) {
            const start = (document as any).startViewTransition;
            if (typeof start !== 'function') return false;

            // The browser snapshots the "before" state synchronously here, calls the callback, and
            // snapshots "after" when the callback's promise resolves. Handing it the promise from
            // invokeMethodAsync is what lets a Blazor render happen between the two snapshots -
            // .NET decides when the DOM is settled, not us.
            const callback = () => butil.utils.dispatch(dotNetRef, 'InvokeViewTransitionUpdate', id);

            let transition: any;
            try {
                // Level 2 takes an options bag with `types`; level 1 takes the callback directly.
                // Passing an object to a level-1 implementation would call it as the callback, so
                // only take that path when types were actually asked for.
                transition = types?.length
                    ? start.call(document, { update: callback, types })
                    : start.call(document, callback);
            } catch {
                return false;
            }

            _transitions[id] = transition;

            // Every phase is a promise, and each is reported separately so .NET can await the one
            // it cares about. A skipped transition rejects `ready` while `finished` still resolves,
            // so the rejection is caught rather than left unhandled.
            transition.ready
                ?.then(() => butil.utils.dispatch(dotNetRef, 'InvokeViewTransitionPhase', id, 'ready', ''))
                .catch((e: any) => butil.utils.dispatch(dotNetRef, 'InvokeViewTransitionPhase', id, 'skipped', e?.message ?? ''));

            transition.finished
                ?.then(() => {
                    delete _transitions[id];
                    butil.utils.dispatch(dotNetRef, 'InvokeViewTransitionPhase', id, 'finished', '');
                })
                .catch((e: any) => {
                    delete _transitions[id];
                    butil.utils.dispatch(dotNetRef, 'InvokeViewTransitionPhase', id, 'failed', e?.message ?? '');
                });

            return true;
        },
        skip(id: string) {
            const transition = _transitions[id];
            if (!transition?.skipTransition) return;
            // Jumps straight to the end state; `finished` still resolves, so the entry is cleaned
            // up by that handler rather than here.
            try { transition.skipTransition(); } catch { /* already finished */ }
        },
        disposeAll() {
            for (const id of Object.keys(_transitions)) {
                butil.viewTransition.skip(id);
                delete _transitions[id];
            }
            for (const id of Object.keys(_pageListeners)) butil.viewTransition.offPageEvent(id);
        }
    };
}(BitButil));
